### PR TITLE
Implemented Kaiser-Bessel derived window

### PR DESCRIPTION
**Pull request by nils-werner**
This PR implements a new window in `scipy.signal`: The Kaiser-Bessel derived window, mainly used with the MDCT transform.

This window is somewhat special as it is only defined for an even number of taps: Requesting a window for an odd number of taps will raise a `ValueError` and the `sym` keyword argument is not defined.

### CODE REVIEW

File: scipy/signal/tests/test_windows.py [hunk 1]
```diff
@@ -213,5 +213,8 @@ def test_windowfunc_basics():
         window(6, *params, sym=False)
 
 
+def test_kaiser_derived():
+    signal.kaiser_derived(6)
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_windows.py]: What is the point of this test?  The only reason to have a test that doesn't check that the return value against a known result is as a regression test for a call that used to, say, generate a seg. fault or incorrectly raise an exception.

File: scipy/signal/windows.py [hunk 2]
```diff
@@ -1065,6 +1066,81 @@ def kaiser(M, beta, sym=True):
...
+    >>> plt.ylabel("Amplitude")
+    >>> plt.xlabel("Sample")
+
+    >>> plt.figure()
+    >>> A = fft(window, 2048) / (len(window)/2.0)
+    >>> freq = np.linspace(-0.5, 0.5, len(A))
+    >>> response = 20 * np.log10(np.abs(fftshift(A / abs(A).max())))
+    >>> plt.plot(freq, response)
+    >>> plt.axis([-0.5, 0.5, -120, 0])
+    >>> plt.title("Frequency response of the kaiser bessel derei window")
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Spelling: "derived"

File: scipy/signal/windows.py [hunk 3]
```diff
@@ -1065,6 +1066,81 @@ def kaiser(M, beta, sym=True):
...
+    >>> plt.ylabel("Normalized magnitude [dB]")
+    >>> plt.xlabel("Normalized frequency [cycles per sample]")
+    >>> plt.show()
+
+    """
+    if M < 1:
+        return np.array([])
+    if M == 1:
+        return np.ones(1, 'd')
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: If the number of coefficient must be even, why is M=1 acceptable?

File: scipy/signal/windows.py [hunk 4]
```diff
@@ -1672,6 +1748,8 @@ def get_window(window, Nx, fftbins=True):
             winfunc = flattop
         elif winstr in ['kaiser', 'ksr']:
             winfunc = kaiser
+        elif winstr in ['kaiser', 'kd', 'kbd']:
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: At this point, `winstr` *can't* be `'kaiser'`.

**Comment by WarrenWeckesser** [comment]
Thanks, the Kaiser-Bessel derived window is a good addition to scipy.

It looks like the implementation is not correct--at least, it doesn't match the definition on the wikipedia page, nor the implementation I found here: www.ee.columbia.edu/~marios/mdct/kbdwin.c

Here's an implementation that matches both those references (and avoids some duplicated computations):

    w = np.zeros(M)
    kaiserw = kaiser(M // 2 + 1, beta)
    csum = np.cumsum(kaiserw)
    halfw = np.sqrt(csum[:-1] / csum[-1])
    w[:M//2] = halfw
    w[-M//2:] = halfw[::-1]

    return w

From what I've read, this window should *not* be normalized to have a maximum of 1.  The values in the window are supposed to satisfy the Princen-Bradley condition: `w[k]**2 +w[k+M//2]**2 = 1`

File: scipy/signal/tests/test_windows.py (hunk 1 above)
Comment (nils-werner) [inline scipy/signal/tests/test_windows.py]: Two lines above it, all other windows are tested like this, too... :-)

File: scipy/signal/windows.py (hunk 3 above)
Comment (nils-werner) [inline scipy/signal/windows.py]: Because falling back to returning 1, like all other windows, seemed sensible to me.

**Comment by nils-werner** [comment]
> Here's an implementation that matches both those references 

Ah, you are right, my denominator was off by one index.

**Comment by nils-werner** [comment]
I've implemented your suggestions and added a simple test for meeting the Princen-Bradley condition and asserting `ValueError` is raised for odd window length.

File: scipy/signal/tests/test_windows.py [hunk 5]
```diff
@@ -213,5 +213,15 @@ def test_windowfunc_basics():
         window(6, *params, sym=False)
 
 
+def test_kaiser_derived():
+    M = 100
+    w = signal.kaiser_derived(M)
+
+    # Test for Princen-Bradley condition
+    assert np.allclose(w[:M//2] ** 2 + w[-M//2:] ** 2, 1.)
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_windows.py]: Don't use a plain python `assert` in tests.  Instead, you can use `assert_allclose(w[:M//2] ** 2 + w[-M//2:] ** 2, 1.)` here.  (`assert_allclose` has already been imported from `numpy.testing`.)

Style nit: I think the general consensus among the scipy developers is to not put spaces around `**`.  So while you're editing this, also remove those spaces.

File: scipy/signal/windows.py (hunk 3 above)
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: I hope they "fall back" to returning 1 because that is the correct (if somewhat degenerate) value to use for the window, and not because they are arbitrarily choosing to do so.

In this case, it is inconsistent to say that the length of the window must be even, but then allow the length to be 1.  How about just removing this check, and letting it raise a ValueError when M is 1?

Comment (nils-werner) [inline scipy/signal/windows.py]: Yup, fine by me.

File: scipy/signal/windows.py [hunk 6]
```diff
@@ -1065,6 +1066,81 @@ def kaiser(M, beta, sym=True):
...
+
+    """
+    if M < 1:
+        return np.array([])
+    if M == 1:
+        return np.ones(1, 'd')
+
+    if M % 2:
+        raise ValueError(
+            "Kaiser Bessel Derived windows are only defined for even number ",
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: There shouldn't be a comma at the end of this line.

File: scipy/signal/tests/test_windows.py [hunk 7]
```diff
@@ -213,5 +213,15 @@ def test_windowfunc_basics():
...
+def test_kaiser_derived():
+    M = 100
+    w = signal.kaiser_derived(M)
+
+    # Test for Princen-Bradley condition
+    assert_allclose(w[:M//2]**2 + w[-M//2:]**2, 1.)
+
+    # Assert ValueError for odd window length
+    assert_raises(ValueError, signal.kaiser_derived, M + 1)
+
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_windows.py]: Style nit: There should be one more blank line between here (between the end of the function definition and the `if` statement`).

**Comment by WarrenWeckesser** [comment]
API question: The `kaiser` window doesn't provide a default value for `beta`.  Do you have a compelling reason for the default value of `beta=4` in the `kaiser_derived` function?  If not (i.e. if `4` was picked arbitrarily), then don't provide a default.

**Comment by WarrenWeckesser** [comment]
You should revert the commit fddef12.  The window coefficients are not normalized to have a maximum of 1. If they were, the Princen-Bradley condition would not be satisfied.  E.g.

```
In [37]: w = kaiser_derived(32, 1.5)

In [38]: w.max()
Out[38]: 0.9786654579117483

In [39]: w[:16]**2 + w[16:]**2
Out[39]: 
array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.])
```

**Comment by nils-werner** [comment]
`beta=4.` is used in some audio coders, so I just used that for my initial experiments. You are right though, it shouldn't have a default value.

**Comment by rgommers** [comment]
This looks about ready to go in, would be good to get it merged before the release split.

Regarding the docstring, it would be useful to add a few sentences about this window's use with the modified discrete cosine transform and that it is commonly used in the audio field.

Would also be useful to squash all commits, and reformat the commit message to conform to the standard format (http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html#writing-the-commit-message). So something like:

    ENH: add a Kaiser-Bessel derived window function to signal

**Comment by endolith** [comment]
Needs to be added to get_window docstring Notes section

**Comment by nils-werner** [comment]
OK, I've implemented your last remarks and rebased the whole PR to sit ontop `master`.

However I noticed my implementation wasn't compatible with `get_window` so I've had to [change it a bit further](https://github.com/nils-werner/scipy/commit/be4f16abae42832d02a17659ea62a604837bdc8b). Can somebody look at it, especially how I am treating the `sym` parameter?

**Comment by nils-werner** [comment]
Any updates on this?

**Comment by WarrenWeckesser** [comment]
I'll take another look this week.

**Comment by endolith** [comment]
Instead of having a dummy `sym` parameter, it would probably be better to just special-case this inside of `get_window`.

```
if win_func == kaiser_derived:
    params = (Nx,) + args
else:
    params = (Nx,) + args + (sym,)
```

Something like that?

File: scipy/signal/tests/test_windows.py [hunk 8]
```diff
@@ -213,14 +213,21 @@ def test_windowfunc_basics():
         window(6, *params, sym=False)
 
 
-def test_needs_params():
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_windows.py]: You have *replaced* this test with `test_kaiser_derived()`.  `test_needs_params` is actually a test of the `get_window` function, and I don't think it should be removed. If there is a good reason for removing it, that should be handled in a separate PR.

File: scipy/signal/windows.py [hunk 9]
```diff
@@ -1065,6 +1066,94 @@ def kaiser(M, beta, sym=True):
     return w
 
 
+def kaiser_derived(M, beta, sym=True):
+    """ Return a Kaiser-Bessel derived window.
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Minor quibble: Don't include the extra space after the opening triple quotes.

File: scipy/signal/windows.py [hunk 10]
```diff
@@ -1065,6 +1066,94 @@ def kaiser(M, beta, sym=True):
...
+
+    Notes
+    -----
+    This window is only defined for an even number of taps.
+
+    It is designed to be suitable for use with the modified discrete cosine
+    transform (MDCT) and is mainly used in audio signal processing and
+    audio coding.
+
+    .. versionadded:: 0.16.0
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Now version 0.18.0 (sorry for the long delays!)

File: scipy/signal/windows.py [hunk 11]
```diff
@@ -1065,6 +1066,94 @@ def kaiser(M, beta, sym=True):
...
+    >>> plt.ylabel("Normalized magnitude [dB]")
+    >>> plt.xlabel("Normalized frequency [cycles per sample]")
+    >>> plt.show()
+
+    """
+    if not sym:
+        raise ValueError(
+            "Kaiser Bessel Derived windows are only defined for symmetric "
+            "shapes"
+        )
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Move this single paren. to the end of the previous line.  (There is an example of this in the PEP 8 document: https://www.python.org/dev/peps/pep-0008/#indentation)

File: scipy/signal/windows.py [hunk 12]
```diff
@@ -1065,6 +1066,94 @@ def kaiser(M, beta, sym=True):
...
+
+    if M < 1:
+        return np.array([])
+
+    if M % 2:
+        raise ValueError(
+            "Kaiser Bessel Derived windows are only defined for even number "
+            "of taps"
+        )
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Same comment here--move the closing paren. to the end of the previous line.

File: scipy/signal/windows.py (hunk 11 above)
Comment (nils-werner) [inline scipy/signal/windows.py]: You mean like

```python
raise ValueError(
    "Kaiser Bessel Derived windows are only defined for symmetric "
    "shapes")
```
?

Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: Yes.

**Comment by WarrenWeckesser** [comment]
A couple more straightforward "to do" items:
* Add `kaiser_derived` to the "Windows" section of the module docstring in `scipy/signal/__init__.py`.
* The function needs tests that check actual values.  Currently the tests only check for consistency with the Princen-Bradley conditions.  I have two independent codes (one octave and one C) that give the following values for half the window when `beta` is `np.pi/2`:

        M = 2:  sqrt(2)/2
        M = 4:  0.518562710536, 0.855039598640
        M = 6:  0.436168993154, 0.707106781187, 0.899864772847

**Comment by WarrenWeckesser** [comment]
@endolith wrote:
> Instead of having a dummy sym parameter, it would probably be better to just special-case this inside of get_window.

But that means `get_window`'s argument `fftbins` is silently ignored for the `kaiser_derived` window.

I don't have a deep understanding of the use-cases of the Kaiser-Bessel derived window.  If it turns out there is a useful definition of the window for an odd number of coefficients, then it is easy to enable this later (and then `kaiser_derived(2*n, sym=False)` can be implemented as `kaiser_derived(2*n+1)[:-1]`).

File: scipy/signal/windows.py [hunk 13]
```diff
@@ -1065,6 +1066,94 @@ def kaiser(M, beta, sym=True):
...
+
+    Examples
+    --------
+    Plot the window and its frequency response:
+
+    >>> from scipy import signal
+    >>> from scipy.fftpack import fft, fftshift
+    >>> import matplotlib.pyplot as plt
+
+    >>> window = signal.kaiser_derived(50)
```
Comment (WarrenWeckesser) [inline scipy/signal/windows.py]: `kaiser_derived` needs a second argument.

**Comment by WarrenWeckesser** [comment]
My final request (really!) is to squash the commits into a single commit, and add our standard prefix `ENH: signal:` to the commit message.  E.g. `ENH: signal: Implemented Kaiser-Bessel derived window`.

**Comment by endolith** [comment]
> But that means get_window's argument fftbins is silently ignored for the kaiser_derived window.

True.

```
if win_func == kaiser_derived:
    if fftbins:
        params = (Nx,) + args
    else:
        raise ValueError "KBD is undefined for fftbins False"
else:
    params = (Nx,) + args + (sym,)
```

> I don't have a deep understanding of the use-cases of the Kaiser-Bessel derived window.

Well it's specifically for MDCT and needs the fftbins=True symmetry for that purpose, so I don't think it will ever need fftbins=False or odd-length.

> A related window function is the Kaiser-Bessel derived (KBD) window, which is designed to be suitable for use with the modified discrete cosine transform (MDCT). ... The KBD window is also symmetric in the proper manner for the MDCT: dn = d2M−1−n.

**Comment by nils-werner** [comment]
> squash the commits into a single commit

Yes, I will obviously do that once everything is cleared up.

**Comment by WarrenWeckesser** [comment]
> Yes, I will obviously do that once everything is cleared up.

Great, thanks.

I'm OK with @endolith's suggestion.  Then `kaiser_derived` only does what it is designed to do, and the convenience function `get_window` takes responsibility for dealing with the small variations in the APIs of the various window function.  What do you think?

**Comment by nils-werner** [comment]
Yes, I can do that.

However I disagree with `get_window` needing to deal with API variations, so I've also created #5744. :-)

**Comment by Eric89GXL** [comment]
This needs a rebase and it looks like this still needs work, so I'm removing the 0.18 label. Let me know if I'm misreading things.

**Comment by rgommers** [comment]
scipy-dev thread: https://mail.scipy.org/pipermail/scipy-dev/2017-January/021639.html

**Comment by larsoner** [comment]
@nils-werner any interest in coming back to this now that `scipy/signal/windows/` exists and has its own namespace?

**Comment by nils-werner** [comment]
Yes, I would like to give this another go.

**Comment by rgommers** [comment]
This PR is continued in gh-15310, so let's close this one. Thanks so far everyone.
